### PR TITLE
[MODFIN-346] - Implement unrelease encumbrance endpoint

### PR DESCRIPTION
**Pull request by azizbekxm**
## Purpose

https://issues.folio.org/browse/MODFIN-346
## Approach

  Endpoint will be implemented
  Unit tests were written
## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?
- Did you modify code to call some additional endpoints?
  - [ ] If so, do you check that necessary module permission added in ModuleDescriptor-template.yaml file?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -1360,6 +1370,11 @@
       "displayName" : "Release encumbrance",
       "description" : "Release any remaining money encumbered back to the budget's available pool"
     },
+    {
+      "permissionName" :  "finance.unrelease-encumbrance.item.post",
+      "displayName" : "UnRelease encumbrance",
+      "description" : "Status of released encumbrance back to unreleased status"
```
Comment (imerabishvili) [inline descriptors/ModuleDescriptor-template.json:1376]: Maybe `Move status` or `Update status`?

File: src/main/java/org/folio/rest/impl/EncumbranceApi.java [hunk 2]
```diff
@@ -36,4 +37,12 @@ public void postFinanceReleaseEncumbranceById(String id, Map<String, String> oka
       .onFailure(fail -> handleErrorResponse(asyncResultHandler, fail));
   }
 
+  @Override
+  @Validate
+  public void postFinanceUnreleaseEncumbranceById(String id, Map<String, String> okapiHeaders,
```
Comment (imerabishvili) [inline src/main/java/org/folio/rest/impl/EncumbranceApi.java:42]: Will repeated call throw an exception?

File: src/main/java/org/folio/services/transactions/CommonTransactionService.java [hunk 3]
```diff
@@ -103,6 +103,25 @@ public Future<Void> releaseTransaction(Transaction transaction, RequestContext r
...
+      .compose(transaction -> unreleaseTransaction(transaction, requestContext));
+  }
+
+  public Future<Void> unreleaseTransaction(Transaction transaction, RequestContext requestContext) {
+    logger.info("Start unreleasing transaction {}", transaction.getId()) ;
+
+    validateTransactionType(transaction, Transaction.TransactionType.ENCUMBRANCE);
+
+    if (transaction.getEncumbrance().getStatus() != Encumbrance.Status.RELEASED) {
+      return succeededFuture(null);
```
Comment (imerabishvili) [inline src/main/java/org/folio/services/transactions/CommonTransactionService.java:117]: Maybe we need to fail because it's called in `POST` method

**Review by Abdulkhakimov — APPROVED** [review]

File: descriptors/ModuleDescriptor-template.json [hunk 4]
```diff
@@ -722,6 +722,16 @@
             "finance-storage.transactions.item.put",
             "finance-storage.order-transaction-summaries.item.put"
           ]
+        },
+        {
+          "method": ["POST"],
+          "pathPattern": "/finance/unrelease-encumbrance/{id}",
+          "permissionsRequired": ["finance.unrelease-encumbrance.item.post"],
```
Comment (SerhiiNosko) [inline descriptors/ModuleDescriptor-template.json:729]: please add this permission to finance.all as we did for finance.release-encumbrance.item.post

File: src/main/java/org/folio/services/transactions/CommonTransactionService.java (hunk 3 above)
Comment (SerhiiNosko) [reply src/main/java/org/folio/services/transactions/CommonTransactionService.java:117]: we need to support consistency with contract of release endpoint, if we going to do change - need to apply the same in that endpoint as well, but as we for me can remain as is

File: src/main/java/org/folio/rest/impl/EncumbranceApi.java (hunk 2 above)
Comment (azizbekxm) [reply src/main/java/org/folio/rest/impl/EncumbranceApi.java:42]: Probably no, I used approach from release encumbrance feature

File: descriptors/ModuleDescriptor-template.json (hunk 4 above)
Comment (azizbekxm) [reply descriptors/ModuleDescriptor-template.json:729]: Added

File: descriptors/ModuleDescriptor-template.json (hunk 1 above)
Comment (azizbekxm) [reply descriptors/ModuleDescriptor-template.json]: Thanks, Updated

**Review by SerhiiNosko — APPROVED** [review]

**Review by imerabishvili — APPROVED** [review]

**Pull request merged**